### PR TITLE
feat(h1): add server header read timeout

### LINKED ISSUE #2457: Defends against slow HTTP attacks.
I think hyper needs to add capabilities such as client_header_timeout and client_body_timeout to defend against slow attacks.

The Nginx defense method is as follows:
https://www.nginx.com/blog/mitigating-ddos-attacks-with-nginx-and-nginx-plus/#client_header_timeout

### PR DESCRIPTION
**Pull request by paolobarbolini**
Implements a timeout for reading the request header. The timer starts either when the connection is established (first request) or when the first byte is read from the header (subsequent requests reusing the same connection). The entire header must be read before the specified timeout is elapsed.

This is the first step towards protecting from Slowloris types of attacks

Supersedes #2661 by fixing https://github.com/hyperium/hyper/pull/2661#discussion_r728809455

Partially fixes #2457

Closes #2661

### CODE REVIEW

**Comment by paolobarbolini** [comment]
Resolved the merge conflicts after the last PR was merged.

Ping @seanmonstar. We're waiting for a code review :smiley:

**Comment by silence-coding** [comment]
@paolobarbolini  From the header_read_timeout_no_write test case, it seems that once the connection is established, the timeout mechanism is started. Is that correct? I think this implementation is lacking, which means that the client's connection pool cannot establish connections in advance, and the client's keepAlive configuration is broken and does not keep alive as expected.

**Comment by paolobarbolini** [comment]
> @paolobarbolini From the header_read_timeout_no_write test case, it seems that once the connection is established, the timeout mechanism is started. Is that correct? I think this implementation is lacking, which means that the client's connection pool cannot establish connections in advance, and the client's keepAlive configuration is broken and does not keep alive as expected.

Right, I didn't consider connection pools starting connections in advance. Fixed

File: Cargo.toml
Comment (seanmonstar) [inline Cargo.toml:89]: In order to keep the runtime stuff optional, I think this should be part of the `runtime` feature, similar to the http2 keep-alive code as well.

File: src/error.rs
Comment (seanmonstar) [inline src/error.rs:109]: The classification of "user" in hyper has so far meant "user of the library", so the programmer writing some code that uses hyper. I wouldn't consider this a "user error", just probably a variant on the main enum.

File: src/proto/h1/role.rs
Comment (seanmonstar) [inline src/proto/h1/role.rs:76]: Is there a way we can either:

- Reuse the `Sleep` future, by calling `reset` with a new deadline.
- Reuse the `Box`, by setting a new one instead of clearing the `Box`.

Either of those would cut down on having to make an extra allocation per request, and we can just have 1 per connection.

File: src/server/server.rs
Comment (seanmonstar) [inline src/server/server.rs:311]: Should we make the argument be an `impl Into<Option<Duration>>`, so people can explicitly turn it off if they want? Is that ever a good idea?

**Review by seanmonstar — CHANGES_REQUESTED** [review]
Sorry for the lag, thanks for writing this! It's very thorough, and I especially appreciate that there's tests so we don't accidentally break it in the future ❤️ 

I left a few comments inline.

File: src/proto/h1/role.rs
Comment (paolobarbolini) [reply src/proto/h1/role.rs]: I ended up reusing the same `Sleep`

File: src/server/server.rs
Comment (paolobarbolini) [reply src/server/server.rs:318]: I'm not sure. The default value is already `None`, so staying with the style of the other methods never setting it should be the way of disabling it?

File: src/server/conn.rs
Comment (seanmonstar) [inline src/server/conn.rs:379]:
```suggestion
    /// transmit the entire header within this time, the connection is closed.
```

File: src/server/server.rs
Comment (seanmonstar) [inline src/server/server.rs:313]:
```suggestion
    /// transmit the entire header within this time, the connection is closed.
```

**Review by seanmonstar — APPROVED** [review]

**Comment by NeoLegends** [comment]
Great work here! This PR leaves me wondering though, won't hyper also need this for h2 and h3, or are there other mechanisms in place there that prevent slow loris attacks?

**Comment by DFINITYManu** [comment]
Valiant effort, but still there is no way to set a timeout for clients that connect but send no data (by experimentation I discovered that the timeout does not fire if I just telnet into the TCP port and hang on there).  Which means it's trivial to gum up a hyper server just by sending a lot of connections and nothing else.

**Pull request merged**
